### PR TITLE
Implement RtlGetLengthWithoutLastFullDosOrNtPathElement and RtlpApplyLengthFunction

### PR DESCRIPTION
**Pull request by learn-more**
JIRA issue: [CORE-11990](https://jira.reactos.org/browse/CORE-11990)

### CODE REVIEW

File: sdk/lib/rtl/path.c [hunk 1]
```diff
@@ -446,20 +446,144 @@ NTAPI
...
+        String = &((PRTL_UNICODE_STRING_BUFFER)UnicodeStringOrUnicodeStringBuffer)->String;
+    }
+    else
+    {
+        DPRINT1("ERROR: Type = %u\n", Type);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    Length = 0;
+    Status = LengthFunction(0, String, &Length);
```
Comment (HBelusca) [inline sdk/lib/rtl/path.c:482]: The Rtlp function doesn't check for LengthFunction being the NULL pointer?

File: sdk/lib/rtl/path.c [hunk 2]
```diff
@@ -446,20 +446,144 @@ NTAPI
...
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    if ((Path->Length / sizeof(WCHAR)) == 0)
+    {
+        /* Nothing to do here */
+        return STATUS_SUCCESS;
+    }
+
+
```
Comment (HBelusca) [inline sdk/lib/rtl/path.c:524]:
```suggestion
```

File: modules/rostests/apitests/ntdll/RtlGetLengthWithoutLastFullDosOrNtPathElement.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * PURPOSE:     Test for RtlGetLengthWithoutLastFullDosOrNtPathElement
+ * COPYRIGHT:   Copyright 2021 Mark Jansen <mark.jansen@reactos.org>
+ */
+
+#include "precomp.h"
+
+
+NTSTATUS
+NTAPI
+RtlGetLengthWithoutLastFullDosOrNtPathElement(
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/RtlGetLengthWithoutLastFullDosOrNtPathElement.c:13]: This function (and the other one in the other test) may be added to the NDK 🤔

File: sdk/lib/rtl/path.c [hunk 4]
```diff
@@ -446,20 +446,144 @@ NTAPI
...
+    {
+        String->Buffer[String->Length / 2] = UNICODE_NULL;
+    }
+
+    return STATUS_SUCCESS;
 }
 
 NTSTATUS
 NTAPI
 RtlGetLengthWithoutLastFullDosOrNtPathElement(IN ULONG Flags,
```
Comment (HBelusca) [inline sdk/lib/rtl/path.c:501]: What about mentioning (in the commit log) that a documentation of this (undocumented) function can be found at:
http://undoc.airesoft.co.uk/ntdll.dll/RtlGetLengthWithoutLastFullDosOrNtPathElement.php ?

File: sdk/lib/rtl/path.c (hunk 1 above)
Comment (learn-more) [reply sdk/lib/rtl/path.c:482]: it does, guess I crashed it on something else.

File: sdk/lib/rtl/path.c (hunk 4 above)
Comment (learn-more) [reply sdk/lib/rtl/path.c:496]: It's mentioned in the apitest.

Comment (HBelusca) [reply sdk/lib/rtl/path.c:496]: Yes it's basically hidden inside. That's why I didn't see it.

Comment (learn-more) [reply sdk/lib/rtl/path.c:496]: We don't generally litter our source tree with comments to documents, unless there is something special there.

Comment (HBelusca) [reply sdk/lib/rtl/path.c:496]: not the source tree. The commit log.

File: modules/rostests/apitests/ntdll/RtlGetLengthWithoutLastFullDosOrNtPathElement.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    Length = 333;
+    Status = RtlGetLengthWithoutLastFullDosOrNtPathElement(0, NULL, &Length);
+    ok_int(Length, 0);
+    ok_hex(Status, STATUS_INVALID_PARAMETER);
+
+    Status = RtlGetLengthWithoutLastFullDosOrNtPathElement(0, &Dum, NULL);
+    ok_hex(Status, STATUS_INVALID_PARAMETER);
+
+    for (ULONG n = 0; n < 32; ++n)
```
Comment (HBelusca) [inline modules/rostests/apitests/ntdll/RtlGetLengthWithoutLastFullDosOrNtPathElement.c:90]: As done in the other test, you can put the `ULONG n` in the list of variables above, and then directly use it there and in the other for-loop without re-declaring it.

File: sdk/lib/rtl/path.c [hunk 6]
```diff
@@ -446,20 +446,139 @@ NTAPI
...
+        return Status;
+
+    if (Length > UNICODE_STRING_MAX_CHARS)
+        return STATUS_NAME_TOO_LONG;
+
+    String->Length = (USHORT)(Length * sizeof(WCHAR));
+
+    if (Type == sizeof(RTL_UNICODE_STRING_BUFFER))
+    {
+        String->Buffer[String->Length / 2] = UNICODE_NULL;
```
Comment (HBelusca) [inline sdk/lib/rtl/path.c:488]:
```suggestion
        String->Buffer[String->Length / sizeof(WCHAR)] = UNICODE_NULL;
```

Comment (HBelusca) [reply sdk/lib/rtl/path.c:488]: or use `Length` ? (since we know the ratio of `String->Length / sizeof(WCHAR)` from the line 484 above) 🤔

**Pull request merged**